File: app/schemas/requirement.py
"""
Cross-Vertical Requirements & Messages Pydantic Schemas
"""

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field
from app.models.requirement import RequirementPriority, RequirementStatus


class RequirementMessageCreate(BaseModel):
    content: str = Field(..., min_length=1, max_length=5000, example="Pitch preparation equipment dispatched.")


class RequirementMessageResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    requirement_id: UUID
    author_id: UUID
    author_username: Optional[str] = None
    author_full_name: Optional[str] = None
    content: str
    created_at: datetime


class RequirementBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=255, example="Request 4 Goal Net Sets for Field B")
    description: str = Field(..., min_length=1, max_length=5000)
    priority: RequirementPriority = RequirementPriority.MEDIUM
    deadline: Optional[datetime] = None
    remarks: Optional[str] = None
    reference_link: Optional[str] = Field(None, max_length=1024, description="Optional reference link (Google Drive, doc, asset, etc.)")


class RequirementCreate(RequirementBase):
    requesting_vertical_id: Optional[UUID] = None
    target_vertical_id: Optional[UUID] = None
    assignee_id: Optional[UUID] = None
    event_id: Optional[UUID] = None


class RequirementUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, min_length=1, max_length=5000)
    priority: Optional[RequirementPriority] = None
    deadline: Optional[datetime] = None
    remarks: Optional[str] = None


class RequirementAssignRequest(BaseModel):
    assignee_id: Optional[UUID] = None


class RequirementForwardRequest(BaseModel):
    target_user_id: Optional[UUID] = Field(None, description="Target user / POC UUID")
    target_vertical_id: Optional[UUID] = Field(None, description="Target vertical UUID if forwarding to vertical")
    reason: str = Field(..., min_length=2, max_length=2000, description="Reason for forwarding internally")


class RequirementEscalateRequest(BaseModel):
    escalated_to_id: UUID = Field(..., description="Target authority/supervisor user UUID")
    reason: str = Field(..., min_length=2, max_length=2000, description="Detailed rationale for escalation")


class RequirementResolveEscalationRequest(BaseModel):
    resolution_notes: str = Field(..., min_length=2, max_length=2000, description="Notes detailing how the escalation was resolved")


class RequirementTransitionRequest(BaseModel):
    status: RequirementStatus
    remarks: Optional[str] = None


class RequirementResponse(RequirementBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    event_id: Optional[UUID] = None
    event_name: Optional[str] = None
    responsible_poc_id: Optional[UUID] = None
    responsible_poc_username: Optional[str] = None
    responsible_poc_full_name: Optional[str] = None
    requesting_vertical_id: Optional[UUID] = None
    requesting_vertical_name: Optional[str] = None
    target_vertical_id: Optional[UUID] = None
    target_vertical_name: Optional[str] = None
    requester_id: UUID
    requester_username: Optional[str] = None
    requester_full_name: Optional[str] = None
    assignee_id: Optional[UUID] = None
    assignee_username: Optional[str] = None
    assignee_full_name: Optional[str] = None
    status: RequirementStatus
    forward_history: List[Dict[str, Any]] = []

    # Escalation fields
    is_escalated: bool = False
    escalated_to_id: Optional[UUID] = None
    escalated_to_username: Optional[str] = None
    escalated_to_full_name: Optional[str] = None
    escalated_by_id: Optional[UUID] = None
    escalated_by_username: Optional[str] = None
    escalated_by_full_name: Optional[str] = None
    escalated_at: Optional[datetime] = None
    escalation_reason: Optional[str] = None
    escalation_status: Optional[str] = None
    escalation_resolved_at: Optional[datetime] = None
    escalation_resolved_by_id: Optional[UUID] = None
    escalation_resolution_notes: Optional[str] = None

    created_at: datetime
    updated_at: datetime
    messages_count: int = 0


class RequirementListResponse(BaseModel):
    total: int
    items: List[RequirementResponse]

File: frontend/types/requirement.ts
/**
 * Cross-Vertical Requirements TypeScript Interfaces
 * Matches backend schemas and database models exactly.
 */

export type RequirementPriority = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export type RequirementStatus =
  | 'RAISED'
  | 'ACKNOWLEDGED'
  | 'IN_PROGRESS'
  | 'AWAITING_INFO'
  | 'FORWARDED'
  | 'ESCALATED'
  | 'RESOLVED'
  | 'CLOSED'
  | 'OPEN'
  | 'ASSIGNED'
  | 'BLOCKED'
  | 'COMPLETED'
  | 'CANCELLED'
  | 'REJECTED';

export interface ForwardHistoryItem {
  forwarded_by_id: string;
  forwarded_by_name: string;
  forwarded_to_id: string;
  forwarded_to_name: string;
  forwarded_to_type: 'USER' | 'VERTICAL';
  reason: string;
  timestamp: string;
}

export interface RequirementMessage {
  id: string;
  requirement_id: string;
  author_id: string;
  author_username?: string;
  author_full_name?: string;
  content: string;
  created_at: string;
}

export interface RequirementResponse {
  id: string;
  title: string;
  description: string;
  event_id?: string;
  event_name?: string;
  responsible_poc_id?: string;
  responsible_poc_username?: string;
  responsible_poc_full_name?: string;
  requesting_vertical_id?: string;
  requesting_vertical_name?: string;
  target_vertical_id?: string;
  target_vertical_name?: string;
  requester_id: string;
  requester_username?: string;
  requester_full_name?: string;
  assignee_id?: string;
  assignee_username?: string;
  assignee_full_name?: string;
  priority: RequirementPriority;
  status: RequirementStatus;
  deadline?: string;
  remarks?: string;
  reference_link?: string;
  forward_history: ForwardHistoryItem[];

  // Escalation fields
  is_escalated: boolean;
  escalated_to_id?: string;
  escalated_to_username?: string;
  escalated_to_full_name?: string;
  escalated_by_id?: string;
  escalated_by_username?: string;
  escalated_by_full_name?: string;
  escalated_at?: string;
  escalation_reason?: string;
  escalation_status?: string;
  escalation_resolved_at?: string;
  escalation_resolved_by_id?: string;
  escalation_resolution_notes?: string;

  created_at: string;
  updated_at: string;
  messages_count: number;
}

export interface RequirementListResponse {
  total: number;
  items: RequirementResponse[];
}

export interface RequirementCreate {
  title: string;
  description: string;
  priority?: RequirementPriority;
  deadline?: string;
  remarks?: string;
  reference_link?: string;
  requesting_vertical_id?: string;
  target_vertical_id?: string;
  assignee_id?: string;
  event_id?: string;
}

export interface RequirementUpdate {
  title?: string;
  description?: string;
  priority?: RequirementPriority;
  deadline?: string;
  remarks?: string;
}

export interface RequirementTransitionRequest {
  status: RequirementStatus;
  remarks?: string;
}

export interface RequirementForwardRequest {
  target_user_id?: string;
  target_vertical_id?: string;
  reason: string;
}

export interface RequirementAssignRequest {
  assignee_id?: string;
}

export interface RequirementEscalateRequest {
  escalated_to_id: string;
  reason: string;
}

export interface RequirementResolveEscalationRequest {
  resolution_notes: string;
}
